feat: allow admins to require invites before user accounts can be created

thenanyu opened the pull request:
Adds a new option in security to disable "Allow authorized signups"
<img width="1552" alt="image" src="https://user-images.githubusercontent.com/427579/163516802-625cd496-53d1-458c-a2d6-8f28b8e1b4bd.png">

Adds a general confirmation dialog component used to confirm the decision to disable
<img width="1552" alt="image" src="https://user-images.githubusercontent.com/427579/163516849-ebec82fe-a9c7-48f0-af50-c74248bea674.png">

When users try to join a team but are not invited, they are ultimately presented with this error state. (I think it looks a little bad stylistically, but that's what's there right now)
<img width="1552" alt="image" src="https://user-images.githubusercontent.com/427579/163632594-602533cb-e220-4fc6-94fb-fb88ec67d9d1.png">

thenanyu commented on app/components/Button.tsx:
```diff
@@ -111,7 +111,7 @@ const RealButton = styled.button<{
 
       &:disabled {
-        background: none;
+        background: ${lighten(0.05, props.theme.danger)};
```
this was displaying as white! I'm a bit surprised this wasn't clearly an issue elsewhere but I guess disabled danger buttons are rare enough

tommoor replied:
If something dangerous is disabled then it's usually hidden I suppose!

thenanyu commented on app/components/ConfirmationDialog.tsx:
```diff
@@ -0,0 +1,58 @@
+import { observer } from "mobx-react";
```
A very general confirmation dialog. Good for simple use cases like interrupting a sensitive config change

thenanyu commented on app/components/Modal.tsx:
```diff
@@ -82,9 +82,11 @@ const Modal: React.FC<Props> = ({
+                        <NudeButton onClick={onRequestClose}>
+                          <CloseIcon color="currentColor" />
+                        </NudeButton>
+                      </Text>
```
the extra text wrapper helps vertically center the glyph visually with the first line of text

thenanyu commented on app/menus/CollectionMenu.tsx:
```diff
@@ -143,7 +143,7 @@ function CollectionMenu({
     dialogs.openModal({
       isCentered: true,
-      title: t("Delete collection"),
+      title: t("Are you sure you want to delete this collection?"),
```
changing to long-title format, sorry for the translation churn!

thenanyu commented on app/scenes/Login/Notices.tsx:
```diff
@@ -71,6 +71,12 @@ export default function Notices() {
           admin.
         </NoticeAlert>
       )}
+      {notice === "invite-required" && (
```
the content of this notice originates with 
```
new Error("Invite required");
```
I'm not thrilled about this, as it seems a bit brittle; open to having a more rigid taxonomy for error types

tommoor replied:
You should create an error in `server/errors` and then it'll be easy to use the existing logic…

thenanyu commented on server/middlewares/passport.ts:
```diff
@@ -13,6 +13,10 @@ export default function createMiddleware(providerName: string) {
         if (err) {
+          if (err.message === "Invite required") {
+            return ctx.redirect(`/?notice=invite-required`);
+          }
```
Again, this seems a bit brittle -- open to suggestions

tommoor commented on app/scenes/Login/Notices.tsx:
```diff
@@ -73,8 +73,10 @@ export default function Notices() {
         <NoticeAlert>
-          Sorry, you must first be invited to your team before you can create an
-          account – To request an invite, please contact a team admin.
+          The team you are trying to join requires an invite before you can
```
This is good 👍

tommoor commented on server/errors.ts:
```diff
@@ -20,6 +20,14 @@ export function AuthorizationError(
 }
 
+export function InviteRequiredError(
+  message = "You need an invite join this team"
```
```suggestion
  message = "You need an invite to join this team"
```

tommoor commented on app/scenes/Settings/Security.tsx:
```diff
@@ -38,22 +42,74 @@ function Security() {
-  };
+  const handleDefaultRoleChange = React.useCallback(
+    async (newDefaultRole: string) => {
+      const newData = { ...data, defaultUserRole: newDefaultRole };
+      await saveData(newData);
```
```suggestion
      await saveData({ ...data, defaultUserRole: newDefaultRole });
```

tommoor commented on app/scenes/Settings/Security.tsx:
```diff
@@ -38,22 +42,74 @@ function Security() {
+                await saveData(newData);
+              }}
+              submitText={t("I’m sure — Disable")}
+              savingText={`${t("Disabling…")}`}
```
```suggestion
              savingText={`${t("Disabling")}…`}
```

We've been keeping these out of the translation string for consistency

tommoor commented on app/scenes/Settings/Security.tsx:
```diff
@@ -38,22 +42,74 @@ function Security() {
+              danger
+            >
+              <Trans
+                defaults="Team members will no longer be able to create accounts on their own with <em>{{authMethodSnippet}}</em> and will need to be invited first."
```
```suggestion
                defaults="New account creation using {{ authenticationMethods }} will be disabled. New users will need to be invited."
```

This appears in the translation interface, maybe change to `authenticationMethods` ? Not sure the bolding is needed here as it should read like part of the sentance.

tommoor commented on app/scenes/Settings/Security.tsx:
```diff
@@ -117,6 +173,28 @@ function Security() {
+        name="allowSignups"
+        description={
+          <Trans
+            defaults="Allow authorized <em>{{ authMethodSnippet }}</em> users to create new accounts without first receiving an invite"
```
```suggestion
            defaults="Allow authorized {{ authMethodSnippet }} users to create new accounts without first receiving an invite"
```

thenanyu replied:
I would rather keep the <em>s around authMethodSnippets

tommoor commented on server/commands/userCreator.test.ts:
```diff
@@ -210,4 +210,33 @@ describe("userCreator", () => {
     expect(isNewUser).toEqual(true);
   });
+
+  it.only("should reject an uninvited user when invites are required", async () => {
```
```suggestion
  it("should reject an uninvited user when invites are required", async () => {
```

thenanyu replied:
oops

tommoor commented on app/actions/definitions/documents.tsx:
```diff
@@ -306,17 +312,43 @@ export const createTemplate = createAction({
-          documentId={activeDocumentId}
-          onSubmit={stores.dialogs.closeAllModals}
-        />
+        <ConfirmationDialog
```
I wish this was still in it's own component, ngl.

thenanyu commented:
A couple of final changes and then we can merge
- factored the template confirmation dialog into its own file again but using the new dialog component
- reverted the inline language strings to not affect the translation file keys and updated the values in the english file instead. I don't know if this is over-optimizing but I wasn't comfortable causing so much translation churn with such a small change

gemathus commented:
Hey guys. I'm curious as to why this setting is only available to cloud-hosted deployments. [This is the commit I'm referring to](https://github.com/outline/outline/commit/964b4ef97dd604862b784af789888e124f21ac8f)

I couldn't find a concrete explanation in commits or pull requests.

tommoor commented:
We decided to make it cloud+enterprise only.

gemathus commented:
Got it. Thanks!
